modin/core/dataframe/pandas/partitioning/partition_manager.py

```diff
@@ -1495,3 +1495,68 @@ def rebalance_partitions(cls, partitions):
             lengths = [part.length() for part in new_partitions[:, 0]]
             return new_partitions, lengths
         return partitions, None
+
+    @classmethod
+    def shuffle_partitions(
+        cls, partitions, index, sample_func, pivot_func, split_func, final_shuffle_func
+    ):
+        """
+        Return shuffled partitions.
+
+        Parameters
+        ----------
+        partitions : np.ndarray
+            The 2-d array of partitions to shuffle.
+        index : int
+            The index of partitions corresponding to the partitions that contain the column to sample.
+        sample_func : Callable(pandas.DataFrame) -> np.ndarray
+            Function to sample partitions. Output of this function will be passed to ``pivot_func``
+            to determine pivots.
+        pivot_func : Callable(List[Any]) -> np.ndarray
+            Function to determine pivots from partitions. The pivots determined by this function
+            will be passed to ``split_func`` to split each partition.
+        split_func : Callable(pandas.DataFrame, Any) -> List[pandas.Dataframe]
+            Function that splits a partition based off of the pivots provided. Should return an
+            unpacked list of pandas Dataframes.
+        final_shuffle_func : Callable(pandas.DataFrame) -> pandas.DataFrame
+            Function that shuffles the data within each new partition.
+
+        Returns
+        -------
+        np.ndarray
+            A list of row-partitions that have been shuffled.
+        """
+        # Mask the partition that contains the column that will be sampled.
+        masked_partitions = partitions[:, index]
+        # Sample each partition
+        sample_func = cls.preprocess_func(sample_func)
+        samples = [partition.apply(sample_func) for partition in masked_partitions]
+        # Get each sample to pass in to the pivot function
+        samples = [sample.get() for sample in samples]
+        pivots = pivot_func(samples)
+        # Convert our list of block partitions to row partitions
+        row_partitions = [
+            partition.force_materialization().list_of_block_partitions[0]
+            for partition in cls.row_partitions(partitions)
+        ]
```

Comment:
isn't it going to merge partitions under the hood (and why do we want to merge them in the first place?.. I mean I get that it's much easier that way, but it's probably more efficient to merge only those partitions which contain sort keys, as moving huge chunks to and from an object store is a costly operation)

also it feels slightly wrong to just take `[0]` partition, as I think we were wanting to generalize on that and make list of blocks be a non-single-item thing in the future in some conditions (so it's not future-proof IMHO)

Reply:
This step is converting the partitions to axis partitions. We want to do it this way since we need to send the whole partition to the `split` step as otherwise we wouldn't know how to split the block partitions that don't contain the sort key. We could always do a broadcast of the sort key, but I'm not sure if we want to handle the complexity of doing that.

Reply:
I wonder if we should add a mini-comment explaining this...